### ISSUE TITLE
Switch to Imperative from Model Mapping approach for SQLAlchemy

### ISSUE THREAD

**Issue by hbooth** [usnistgov/dioptra#185]
SQLAlchemy is moving toward a more direct SQL approach for ORM and away from Mapping Python classes.
Switching would allow a looser coupling between the ORM and classes.
Additionally, this would be a good opportunity to consolidate all ORM to a common place in the code.

# Definition of Done
- [ ] Switched all mapping to Direct SQL approach in SQL Alchemy
- [ ] All Queries use new mappings
- [ ] No change in Database
- [ ] Accepted and merged into Dev branch

**Comment by jkglasbrenner** [comment]
Plan: I will show how to do this on the queue endpoint, and then @hbooth and I will collaborate to convert the rest of the endpoints after #161 is completed.

**Comment by jkglasbrenner** [comment]
@hbooth Migrating away from the ORM data models that inherit from `db.Model` to using Python's dataclasses and [SQLAlchemy's Imperative Mapping style](https://docs.sqlalchemy.org/en/14/orm/mapping_styles.html#orm-imperative-mapping) isn't as straightforward as I originally thought, mostly because I hadn't considered how Flask-SQLAlchemy itself works. To use Flask-SQLAlchemy v2.x (we cannot move to v3.x right now because of other dependency constraints), you must use the [Declarative Mapping](https://docs.sqlalchemy.org/en/14/orm/mapping_styles.html#declarative-mapping) style and, in addition, [inherit from its `db.Model` class](https://flask-sqlalchemy.palletsprojects.com/en/2.x/models/). In short, Flask-SQLAlchemy is doing some magic behind-the-scenes that relies on knowing which classes inherit from `db.Model`.

For this reason, the only way that we could move to the imperative mapping style is to stop using Flask-SQLAlchemy as a dependency, which will require more work than it first seemed. I suspect that moving to the imperative mapping style also means we would need to stop using Flask-Migrate, and instead use the alembic tool directly (that part isn't so bad). The main consequence is that decoupling from these two dependencies will likely take longer than we estimated, and that refactoring the code base will require halting further development on the REST API until the refactoring is complete.

On my other project where I managed SQLAlchemy and alembic more directly, I used the ideas in the [Architecture Patterns with Python book (available for free online)](https://www.cosmicpython.com/book/preface.html), specifically the chapters on the Repository Pattern, Unit of Work Pattern, Aggregates and Consistency Boundaries, and Command-Query Responsibility Segregation (CQRS). I also used the injector library (also used in our project) to handle the dependency injection needed to make the above work. We can go this road, and I'm guessing the ideas/patterns I'm referencing are quite familiar to you. I think we will need to spend a little bit of time designing our _Aggregates_ based on the current database schema. Of course, doing this will take additional time, and perhaps we'd rather just put a pin in it for now and focus on moving forward with new features first.

I've pushed the initial start towards the imperative mapping style to the branch refactor-sqlalchemy-models. Basically I just created a `db` folder within the `src/dioptra/restapi` directory and put some initial code in it. It's not currently hooked up to anything, but it can give you an initial idea on how some of this would look.

**Comment by chisholm** [comment]
How about this (using flask-sqlalchemy 2.5.1, and a made-up table):

```python
software_table = db.Table(
    "software",
    db.Model.metadata,
    db.Column("id", db.Integer, primary_key=True),
    db.Column("name", db.String(30)),
    db.Column("os", db.String(30))
)


class Software:
    pass


db.Model.registry.map_imperatively(Software, software_table)
```

Seems like flask-sqlalchemy puts the basic pieces in place to do this, they just don't document this style.  Metadata and registry objects are there to use, just need to dig them out of the declarative base.

**Comment by jkglasbrenner** [comment]
Superseded/closed by #463. Leveraged [Declarative Dataclass Mapping](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping) instead.
